use project_options and cmake v3.21

ClausKlein opened the pull request.

ClausKlein committed:
use project_options and cmake v3.21

ClausKlein committed:
refactory cmake project concept

the all is the developer enabled source dir
the project root is for build and install only
test or all may be used for test coverage
format all files again

ClausKlein committed:
upgrad cmake-format version

back to old test strategie with option

ClausKlein committed:
disable project_options for now

first step must be to fix the doctest linker problems!

ClausKlein commented on test/CMakeLists.txt:
```diff
@@ -16,37 +16,40 @@ include(../cmake/tools.cmake)
+
 add_executable(${PROJECT_NAME} ${sources})
-target_link_libraries(${PROJECT_NAME} doctest::doctest Greeter::Greeter)
-target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_20)
```
check this!

ClausKlein replied:
check this!

ClausKlein commented on test/CMakeLists.txt:
```diff
@@ -16,37 +16,40 @@ include(../cmake/tools.cmake)
-# enable pedantic compiler warnings
-include(../cmake/WarningsAsErrors.cmake)
+  if(ENABLE_TEST_COVERAGE)
+    target_compile_options(greeter PUBLIC -O0 -g -fprofile-arcs -ftest-coverage)
```
this is the old way!
but static configured and difficult to understand!

ClausKlein replied:
this is the old way!
but static configured and difficult to understand!

ClausKlein commented on CMakeLists.txt:
```diff
@@ -33,37 +47,77 @@ CPMAddPackage(
 # Note: for header-only libraries change all PUBLIC flags to INTERFACE and create an interface
 # target: add_library(${PROJECT_NAME} INTERFACE)
 add_library(${PROJECT_NAME} ${headers} ${sources})
+add_library(${PROJECT_NAME}::${PROJECT_NAME} ALIAS ${PROJECT_NAME})
```
should not be needed!
is done by `PackageProject.cmake`

ClausKlein replied:
@aminya May automatically generated by `package_project()` too.
Note: this is what `PackageProject.cmake` do.

ClausKlein commented on CMakeLists.txt:
```diff
@@ -33,37 +47,77 @@ CPMAddPackage(
+  return()
+endif()
+
+if(NOT PROJECT_IS_TOP_LEVEL)
```
**TODO**: should it possible to install developer versions?

It is not a good idea to export `code_coverage` or `sanitizer` compiler options!

ClausKlein replied:
**TODO**: should it possible to install developer versions?

**IMHO**: It is not a good idea to export `code_coverage` or `sanitizer` compiler options!

ClausKlein committed:
fix the doctest linker problems

quickfix only, I will switch to boost::ut

ClausKlein committed:
back to C++20 standard as default

ClausKlein commented on test/CMakeLists.txt:
```diff
@@ -13,40 +13,40 @@ include(../cmake/tools.cmake)
 
-if(TEST_INSTALLED_VERSION)
-  find_package(Greeter 1.0 REQUIRED)
+if(TEST_INSTALLED_VERSION) # FIXME: if(PROJECT_IS_TOP_LEVEL)
```
The main question: is `test` as `top level project` dependent of installed version by default?
In original version **not** and use for `test_coverage`!

ClausKlein committed:
use projcet_options again

prevent cmake-format for package_project()

ClausKlein committed:
update readme and some cleanups

ClausKlein committed:
fix typos in readme and code

ClausKlein committed:
remove unused cmake module

ClausKlein committed:
add package_project cmake-syntax

ClausKlein committed:
extent perlre for -Wformat=2 warnings

ClausKlein committed:
prevent to export pedantic developer mode settings

change install prefix to /usr to prevent install problems

aminya commented on CMakeLists.txt:
```diff
@@ -33,49 +50,86 @@ CPMAddPackage(
+# This overwrites `project_options` and sets `project_warnings`
+# uncomment the options to enable them:
+# cmake-format: off
+dynamic_project_options(
```
I prefer to use `project_options` instead of `dynamic_project_options`. See the readme:
https://github.com/aminya/project_options#usage

ClausKlein replied:
Agreed, for **Ci**!
But for development in directory `all` not. I like the flexibility.

Another option would be a set of useful `cmake presets`.

aminya commented on CMakeLists.txt:
```diff
@@ -1,28 +1,45 @@
+# the same version
+set(CMAKE_CXX_STANDARD 20)
+
+# strongly encouraged to enable this globally to avoid conflicts between -Wpedantic being enabled
+# and -std=c++20 and -std=gnu++20 for example when compiling with PCH enabled
+set(CMAKE_CXX_EXTENSIONS NO)
+
+# to be prepared for clang-tools usage!
+set(CMAKE_EXPORT_COMPILE_COMMANDS YES)
+
```
Already set inside project_options

```suggestion
```

ClausKlein replied:
Yes, but I have to test if it works, may be it must be defined before `project()`

I have to check this.

ClausKlein committed:
add cmake preset support and a default preset

remove not relevant reverences to ModernCppStarter project site

ClausKlein committed:
move the cmake preset to test subdir

this is the main usage example for developers
add some notes to the CMakeLists.txt file about there usage

ClausKlein commented on README.md:
````diff
@@ -44,9 +58,20 @@ This template is the result of learnings from many previous projects and should
+
+```bash
+cd test
+cmake --preset=test
````
@jenisys now we have a preset which works!

ClausKlein replied:
CMAKE_EXPORT_COMPILE_COMMANDS is not needed here, removed.

The other are set to have a defined Standard set for all directories. At least for

ClausKlein commented on CMakeLists.txt:
```diff
@@ -1,28 +1,45 @@
-cmake_minimum_required(VERSION 3.14...3.22)
+#
+# this project is for using as a sub project with add_subdirectory() or with CPMAddPackage()
```
Note this: No tests, only build and install, no developer mode!

ClausKlein commented on test/CMakeLists.txt:
```diff
@@ -1,10 +1,14 @@
-cmake_minimum_required(VERSION 3.14...3.22)
+#
+# This project enables the developer to build, test, install the project.
```
Note this: The developer mode may be enabled, and any other `project_options OPT_....`

ClausKlein commented on all/CMakeLists.txt:
```diff
@@ -1,10 +1,21 @@
 
 project(BuildAll LANGUAGES CXX)
 
+# Note: by default ENABLE_DEVELOPER_MODE is True.
```
this is for working with an IDE ...

ClausKlein committed:
update readme and cmake presets

ClausKlein committed:
fix typo in readme
